### ISSUE TITLE
[Tracking Issue/Discussion] The Ecosystem Guide

### ISSUE THREAD

**Issue by Lokathor** [rust-gamedev/rust-gamedev.github.io#6]
I know that there's an open PR for the start of an ecosystem guide, but since this is one of our goals before the next meeting I wanted to have a general discussion and get people's thoughts about the subject without totally clogging up the review of kvark's specific PR.

The general subject of this issue is _anything_ related to the gamedev-wg doing an ecosystem guide.

Assuming that we do make an ecosystem guide, I feel compelled to ask: What format should such a guide take?

1) Obvious first point: [AreWeGameYet](http://arewegameyet.com/) is _already_ an ecosystem guide.
  * We don't maintain it, though this is not itself a problem because it does get active attention.
  * There's a lot of open **issues** but not many open **PRs**. It seems like _once someone does the PR work_ that PR gets merged fairly promptly, but the owner isn't constantly putting in the new entries themselves.
  * AreWeGameYet is fairly un-opinionated regarding one library or another, it just sorta throws everything into the pile with minimal descriptions and lets readers figure things out for themselves.
2) [The Charter](https://github.com/rust-gamedev/wg/blob/master/charter.md) of this WG _fairly specifically_ says that we should **not** be promoting one library over another.
  * AreWeGameYet fills the space of an un-opinionated list of everything.
  * The only good space after that is to have an opinionated list of everything (and perhaps some things don't even make it onto the list, which is itself an opinion).

Combined, these two items make me want to back up and ask: Do we make an ecosystem guide at all?

* I think that there _should_ be an ecosystem guide that's says a lot more about each particular library in the guide. Like the [Guide To Rust Graphics in 2019](https://wiki.alopex.li/AGuideToRustGraphicsLibraries2019) that @icefoxen wrote. That article is really good, and frank opinions about the fitness for use of major crates is really valuable advice for the community.
* However, as soon as you say anything more than basic facts about each crate you start having to have an opinion about things, which is pretty close to "promoting one library over another", which the charter says for the WG to not do.
* I do recall from seeing some early discussions before the WG was formed that some folks wanted the WG to rally the community around one specific library for each particular programming element. Getting everyone to pour all effort into Amethyst or Nalgebra or whatever library. That clause in the charter seems to be a reaction to that idea, and a lesser statement of just saying "these libraries are actually fit for use for developing games right now [list here]" might be acceptable in the long term.

My conclusion from these facts: Either we have to change the charter at least a bit (not an action to be taken lightly!!) or we can't have an ecosystem guide that is usefully different from AreWeGameYet, and we probably shouldn't do a separate guide at all. Instead, we should just direct people to work on AreWeGameYet.

Thoughts on this? Or thoughts on anything else about the subject of an Ecosystem Guide?

**Comment by Wodann** [comment]
During our biweekly [meeting](rust-gamedev/wg/issues/38) we coined multiple ways of improving AreWeGameYet organisation:

1. Indicate the "liveness": if something is outdated versus recently updated
2. Classify with features that we (the community and the working group) all agree on. Modifying the feature labels would need to be peer-reviewed so that the classification stays objective at large.
3. Give a clue about the tech stack ("main" dependencies) as well as compatibility (e.g. "this works with winit").

Complementary to that we'll create a mapping of the ecosystem similar to Jason Gregory's [runtime game engine architecture map](http://hightalestudios.com/wordpress/wp-content/uploads/2017/03/jaKUP.png). This will allow us to identify missing pieces.

How this will actually take shape is to be decided, but we'd really appreciate AreWeGameYet's input (@doppioslash) in the discussion. Let me coin a first proposal, inspired by the above thread. Feel free to propose additions, improvements, or provide arguments as to why these are terrible ideas.

### 1. Liveness
- Add a flag to indicate whether the crate is actively/passively/not maintained
- Add minimum Rust version to crate overview
- Add date of last update to crate overview
- Add git statistics, such as number of forks, stars, and latest commit (to give an indication of developer activity)
- Allow sorting of crates based on downloads, minimum Rust version, date of last update, and git stats

Caveat: Depending on how much information is added, the UI could become too crowded. As such this might require a redesign.

### 2. Feature comparison
AreWeGameYet already subdivides the ecosystem into 19 different categories. After reviewing the categories to reflect all aspects of a game engine, we'll try to specify a set of feature tags for each of the categories. For every category we will then:

- tag the supported features of each crate
- tag the supported operating systems of each crate
- show a comparison table of the crates' features

The tricky part is that merely "having" a ton of features might not be the best criteria. E.g. for an ECS you might be interested in the performance of a lot of entities vs a lot of components. Does anyone have any good ideas on how to reflect these objectively? Maybe

- Add some kind of quantitative measures in the overview
- Add an *Additional Resources* section that links to in-depth comparisons/blogs

### 3. Compatibility
- Add nightly/stable to crate overview
- Show tech stack by mentioning dependencies one layer down in the ecosystem; e.g. according to Jason Gregory's [runtime game engine architecture map](http://hightalestudios.com/wordpress/wp-content/uploads/2017/03/jaKUP.png). An example would be:

- Animation Crate A, with ragdoll physics (depends on Physics Crate B)
- Physics Crate B (depends on Math Crate C)

This way one can jump through the ecosystem from A to B and from B to C

**Comment by aclysma** [comment]
Personally, I like a lot of the ideas you've proposed. (Although it may be a lot of work!)

> Add a flag to indicate whether the crate is actively/passively/not maintained

I think this would be useful, but it sounds difficult to do in an automated way, and I think it would be liable to rot. Would be really nice to have this if possible though.

> Add an Additional Resources section that links to in-depth comparisons/blogs

Linking to other resources could be a practical way to be helpful while staying unopinionated as a WG. (For example, if there's a good "what are people using for physics" reddit thread and a there are a lot of good responses, that might be handy for someone taking a look around the ecosystem.) There may not be that many resources we can point to just yet, but I could see this improving as the ecosystem matures.

**Comment by Wodann** [comment]
@aclysma I agree that it is a lot of work. I was hoping to prioritise the list based on feedback and implement accordingly.

The actively/passively maintained flags would be set upon insertion of a new crate, as proposed in @kvark 's format. Ideally, maintainers would change the status once they stop maintaining their repo - similar to how they sometimes add a notification in their README. I agree with you that in reality it will probably mean some maintenance work for the WG though. I.e. review the status of inactive repos every month or so.

@doppioslash I realise that there are a lot of proposed ideas. Could you give some insight into what aligns with AreWeGameYet's vision, and what is achievable?

**Comment by icefoxen** [comment]
Crates.io already has a "maintenance status" tag, we could just display that if we trust the owners of the crates to put meaningful stuff there.

**Comment by Wodann** [comment]
@icefoxen That's a good one!

For completeness:
```
# Optional specification of badges to be displayed on crates.io.
#
# - The badges pertaining to build status that are currently available are
#   Appveyor, CircleCI, GitLab, Azure DevOps and TravisCI.
# - Available badges pertaining to code test coverage are Codecov and
#   Coveralls.
# - There are also maintenance-related badges based on isitmaintained.com
#   which state the issue resolution time, percent of open issues, and future
#   maintenance intentions.
#
# If a `repository` key is required, this refers to a repository in
# `user/repo` format.
[badges]
```
